linalg/sundials.hpp

```diff
@@ -43,25 +47,159 @@ namespace mfem
+class SundialsNVector : public Vector
+{
+protected:
+   int own_NVector;
+
+   /// The actual SUNDIALS object
+   N_Vector x;
+
+   friend class SundialsSolver;
+
+   /// Set data and length of internal N_Vector x from 'this'.
+   void _SetNvecDataAndSize_(long glob_size = 0);
+
+   /// Set data and length from the internal N_Vector x.
+   void _SetDataAndSize_();
+
+public:
+   /// Creates an empty SundialsNVector.
+   SundialsNVector();
+
+   /// Creates a SundialsNVector referencing an array of doubles, owned by someone else.
+   /** The pointer @a _data can be NULL. The data array can be replaced later
+       with SetData(). */
+   SundialsNVector(double *_data, int _size);
+
+   /// Creates a SundialsNVector out of a SUNDIALS N_Vector object.
+   /** The N_Vector @a nv must be destroyed outside. */
+   SundialsNVector(N_Vector nv);
+
+#ifdef MFEM_USE_MPI
+   /// Creates an empty SundialsNVector.
+   SundialsNVector(MPI_Comm comm);
+
+   /// Creates a SundialsNVector with the given local and global sizes.
+   SundialsNVector(MPI_Comm comm, int loc_size, long glob_size);
+
+   /// Creates a SundialsNVector referencing an array of doubles, owned by someone else.
+   /** The pointer @a _data can be NULL. The data array can be replaced later
+       with SetData(). */
+   SundialsNVector(MPI_Comm comm, double *_data, int _size, long glob_size);
+
+   /// Creates a SundialsNVector from a HypreParVector.
+   /** Ownership of the data will not change. */
+   SundialsNVector(HypreParVector& vec);
+#endif
+
+   /// Calls SUNDIALS N_VDestroy function if the N_Vector is owned by 'this'.
+   ~SundialsNVector();
+
+   /// Returns the N_Vector_ID for the internal N_Vector.
+   inline N_Vector_ID GetNVectorID() const { return N_VGetVectorID(x); }
+
+   /// Returns the N_Vector_ID for the N_Vector @a _x.
+   inline N_Vector_ID GetNVectorID(N_Vector _x) const { return N_VGetVectorID(_x); }
+
+#ifdef MFEM_USE_MPI
+   /// Returns the MPI commmunicator for the internal N_Vector x.
+   inline MPI_Comm GetComm() const { return *static_cast<MPI_Comm*>(N_VGetCommunicator(x)); }
+
+   /// Returns the MPI global length for the internal N_Vector x.
+   inline long GlobalSize() const { return N_VGetLength(x); }
+#endif
+
+   /// Resize the vector to size @a s.
+   void SetSize(int s, long glob_size = 0);
+
+   /// Set the vector data.
+   /// @warning This method should be called only when OwnsData() is false.
+   void SetData(double *d);
+
+   /// Set the vector data and size.
+   /** The Vector does not assume ownership of the new data. The new size is
+       also used as the new Capacity().
+       @warning This method should be called only when OwnsData() is false. */
+   void SetDataAndSize(double *d, int s, long glob_size = 0);
+
+   /// Reset the Vector to be a reference to a sub-vector of @a base.
+   inline void MakeRef(Vector &base, int offset, int s)
+   {
+      // Ensure that the base is registered/initialized before making an alias
+      base.Read();
```

Comment:
This is not optimal, but see https://github.com/mfem/mfem/issues/1578#issuecomment-649962962.